src/main/xerus/monstercat/api/Playlist.kt

```diff
@@ -0,0 +1,91 @@
+package xerus.monstercat.api
+
+import javafx.collections.FXCollections
+import javafx.collections.ObservableList
+import mu.KotlinLogging
+import xerus.ktutil.javafx.properties.SimpleObservable
+import xerus.ktutil.javafx.properties.bindSoft
+import xerus.monstercat.api.response.Track
+import java.util.*
+import kotlin.collections.ArrayList
+
+object Playlist {
+	val logger = KotlinLogging.logger { }
+	
+	val tracks: ObservableList<Track> = FXCollections.observableArrayList()
+	val history = ArrayDeque<Track>()
+	val currentIndex = SimpleObservable<Int?>(null).apply {
+		bindSoft({
+			tracks.indexOf(Player.activeTrack.value).takeUnless { i -> i == -1 }
+		}, Player.activeTrack, tracks)
+	}
+	
+	var repeat = false
+	var shuffle = false
+	
+	operator fun get(index: Int): Track? = tracks.getOrNull(index)
+	
+	/** This property prevents songs that come from the [history] from being added to it again,
+	 * which would effectively create an infinite loop */
+	var lastPolled: Track? = null
+	
+	fun getPrev(): Track? = history.pollLast()?.also {
+		lastPolled = it
+		logger.debug("Polled $it from History")
+	} ?: currentIndex.value?.let { get(it - 1) }
+	
+	fun getNext() : Track?{
+		return when {
+			shuffle -> nextSongRandom()
+			repeat && (nextSong() == null) -> tracks[0]
+			else -> nextSong()
+		}
+	}
+	
+	fun addNext(track: Track) = tracks.run {
+		remove(track)
+		add(currentIndex.value?.let { it + 1 } ?: 0, track)
+	}
+	
+	fun add(track: Track): Boolean = tracks.run {
```

Comment:
please don't use `run` here, because it requires the reading developer to switch contexts.

Reply:
363824a8dbe724ce172257b20ab8b1548b5378cf

Reply:
You haven't changed anything? `run` and `apply` should be used only if there is a very strong reason.

Reply:
Weird, I changed the wrong function.

Reply:
What do you think of this ? c3065c87d2f4f556cec70ce0840bdbd7f90c855e
Or should I just remove the apply and do tracks.remove; tracks.add ?